jib-core/src/main/java/com/google/cloud/tools/jib/configuration/ContainerConfiguration.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2018 Google LLC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.configuration;
+
+import com.google.cloud.tools.jib.image.json.BuildableManifestTemplate;
+import com.google.cloud.tools.jib.image.json.V22ManifestTemplate;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import javax.annotation.Nullable;
+
+/** Immutable configuration options for the container. */
+public class ContainerConfiguration {
+
+  /** Builder for instantiating a {@link ContainerConfiguration}. */
+  public static class Builder {
+
+    private Instant creationTime = Instant.EPOCH;
+    @Nullable private ImmutableList<String> entrypoint;
+    @Nullable private ImmutableList<String> programArguments;
+    @Nullable private ImmutableMap<String, String> environmentMap;
+    @Nullable private ImmutableList<Port> exposedPorts;
+    private Class<? extends BuildableManifestTemplate> targetFormat = V22ManifestTemplate.class;
+
+    /**
+     * Sets the image creation time.
+     *
+     * @param creationTime the creation time
+     * @return this
+     */
+    public Builder setCreationTime(Instant creationTime) {
+      this.creationTime = creationTime;
+      return this;
+    }
+
+    /**
+     * Sets the commandline arguments for main.
+     *
+     * @param programArguments the list of arguments
+     * @return this
+     */
+    public Builder setProgramArguments(@Nullable List<String> programArguments) {
+      if (programArguments != null) {
+        Preconditions.checkArgument(!programArguments.contains(null));
+        this.programArguments = ImmutableList.copyOf(programArguments);
+      }
+      return this;
+    }
+
+    /**
+     * Sets the container's environment variables, mapping variable name to value.
+     *
+     * @param environmentMap the map
+     * @return this
+     */
+    public Builder setEnvironment(@Nullable Map<String, String> environmentMap) {
+      if (environmentMap == null) {
+        this.environmentMap = null;
+      } else {
+        Preconditions.checkArgument(
+            !Iterables.any(environmentMap.keySet(), Objects::isNull)
+                && !Iterables.any(environmentMap.values(), Objects::isNull));
```

Comment:
Huh, now I see I missed this in the original PR a while ago. This should actually be broken into two `checkArgument()`s, so that we know which one was at fault if these are violated.